### PR TITLE
[confighttp] Change Headers field type to have opaque values

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:**

Change `Headers` field on HTTP client config to have opaque values and transparent keys.

Depends on #6470

**Link to tracking Issue:** Updates #5653

### CODE REVIEW

File: config/confighttp/confighttp.go
Comment (jpkrohling) [inline config/confighttp/confighttp.go:162]: Would it make sense to convert the string headers to opaque headers here?

Comment (mx-psi) [reply config/confighttp/confighttp.go:162]: Sure, it doesn't hurt to use the opaque type there too :)

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [inline config/confighttp/confighttp.go:59]: Implement "Unmarshal" func on the struct, and during that initialize Headers with the values from `OpaqueHeaders`? That way at least the "unmarshalling" is backwards compatible?

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [inline config/confighttp/confighttp.go:64]: Not convinced that I want to endup with this name. What do you think? Is there a way to use the previous name, maybe a few versions plan?

Comment (mx-psi) [reply config/confighttp/confighttp.go:64]: My idea was to follow this plan: https://github.com/open-telemetry/opentelemetry-collector/pull/6470#issuecomment-1332094426 with which we would end up back with `Headers`

File: config/confighttp/confighttp.go
Comment (mx-psi) [reply config/confighttp/confighttp.go:59]: > Implement "Unmarshal" func on the struct

Does this work with structs that are not component config structs? I wasn't aware, I can do that.

Note that with the current code unmarshaling is kept backwards compatible in the sense that for a given confmap we end up building the same client as before. But if we care about people accessing the individual fields we should do the unmarshal thing, yes, will fix that

Comment (bogdandrutu) [reply config/confighttp/confighttp.go:59]: Yes works, since 1-2 versions ago :)

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [reply config/confighttp/confighttp.go:64]: I know that just doing `Headers map[string]configopaque.String` is a "backwards" incompatible change, but it is very limited incompatibility which I believe we can make an exception for to speed this up.

Comment (mx-psi) [reply config/confighttp/confighttp.go:64]: I am fine with that, but I don't think @jpkrohling is, so I am trying to reach a compromise here :)

File: config/confighttp/confighttp.go
Comment (mx-psi) [reply config/confighttp/confighttp.go:59]: I think it does not work for embedded structs? See #6668, will open issue if you confirm this is indeed a bug

Comment (bogdandrutu) [reply config/confighttp/confighttp.go:59]: This is not embedded, but that is a bug indeed. Do you plan to fix it?

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [reply config/confighttp/confighttp.go:64]: Ping @jpkrohling

File: config/confighttp/confighttp.go
Comment (mx-psi) [reply config/confighttp/confighttp.go:59]: > This is not embedded, but that is a bug indeed

I ran into the issue because the client struct is embedded on the OTLP HTTP exporter https://github.com/open-telemetry/opentelemetry-collector/blob/fb0842a5a200ad1bc7475a4f69baa994310792a4/exporter/otlphttpexporter/config.go#L29 and some tests were failing

> Do you plan to fix it?

Not today, but I can do it next week

File: config/confighttp/confighttp.go
Comment (mx-psi) [reply config/confighttp/confighttp.go]: I think the usage outside of contrib is minimal and the error and fix is quite clear (just change the type), so I change to do it in a single step. We can consider case by case on other fields

**Comment by mx-psi** [comment]
I changed this PR to do a breaking change, since the breakage seems limited enough to do in a single step. We will probably need to skip contrib tests and fix things over there in a version bump (I promise to do it ASAP once this is merged within my working hours :))

File: config/confighttp/confighttp.go
Comment (mx-psi) [reply config/confighttp/confighttp.go]: This no longer blocks this PR, so we can continue the discussion on #6671. It sounds like it's not easy to fix though (may need vendoring the Decode function?)

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
